File: Days 21 - 30/Day 25/SimpleRegexImplementation.cpp
#include <iostream>
#include <string>
#include <vector>

bool MatchSimpleRegex(const std::string& regex, const std::string& text)
{
	std::vector<std::vector<bool>> matches(text.length() + 1, std::vector<bool>(regex.length() + 1));
	matches[0][0] = true;

	static constexpr char SINGLE_CHARACTER = '.';
	static constexpr char MULTIPLE_CHARACTERS = '*';

	for (unsigned int i = 1; i < matches[0].size(); i++)
	{
		if (regex[i - 1] == MULTIPLE_CHARACTERS)
		{
			matches[0][i] = matches[0][i - 2];
		}
	}

	for (unsigned int i = 1; i < matches.size(); i++)
	{
		for (unsigned int j = 1; j < matches[i].size(); j++)
		{
			if ((regex[j - 1] == SINGLE_CHARACTER) || (regex[j - 1] == text[i - 1]))
			{
				matches[i][j] = matches[i - 1][j - 1];
			}
			else if (regex[j - 1] == MULTIPLE_CHARACTERS)
			{
				matches[i][j] = matches[i][j - 2];

				if ((regex[j - 2] == SINGLE_CHARACTER) || (regex[j - 2] == text[i - 1]))
				{
					matches[i][j] = matches[i][j] || matches[i - 1][j];
				}
			}
			else
			{
				matches[i][j] = false;
			}
		}
	}

	return matches[text.length()][regex.length()];
}

int main(int argc, const char* argv[])
{
	std::cout << std::boolalpha;

	std::cout << MatchSimpleRegex(".*at", "chat") << "\n";
	std::cout << MatchSimpleRegex(".*at", "chats") << "\n";

	std::cout << MatchSimpleRegex("ra.", "ray") << "\n";
	std::cout << MatchSimpleRegex("ra.", "raymond") << "\n";

	std::cout << MatchSimpleRegex(".*", "aaaaaaaaaaaaaa") << "\n";
	std::cout << MatchSimpleRegex(".", "aaaaaaaaaaaaaa") << "\n";
	
	std::cin.get();

	return 0;
}

File: Days 021 - 030/Day 25/SimpleRegexImplementation.cs
using System;

namespace DailyCodingProblem
{
	internal class Day25
	{
		private static int Main(string[] args)
		{
			Console.WriteLine(MatchSimpleRegex(".*at", "chat"));
			Console.WriteLine(MatchSimpleRegex(".*at", "chats"));

			Console.WriteLine(MatchSimpleRegex("ra.", "ray"));
			Console.WriteLine(MatchSimpleRegex("ra.", "raymond"));

			Console.WriteLine(MatchSimpleRegex(".*", "aaaaaaaaaaaaaaaaaa"));
			Console.WriteLine(MatchSimpleRegex(".", "aaaaaaaaaaaaaaaaaa"));

			Console.ReadLine();

			return 0;
		}

		private static bool MatchSimpleRegex(string regex, string text)
		{
			bool[,] matches = new bool[text.Length + 1, regex.Length + 1];
			matches[0, 0] = true;

			const char SingleCharacter = '.';
			const char MultipleCharacters = '*';

			for (int i = 1; i < matches.GetLength(1); i++)
			{
				if (regex[i - 1] == MultipleCharacters)
				{
					matches[0, i] = matches[0, i - 2];
				}
			}

			for (int i = 1; i < matches.GetLength(0); i++)
			{
				for (int j = 1; j < matches.GetLength(1); j++)
				{
					if (regex[j - 1] == SingleCharacter || regex[j - 1] == text[i - 1])
					{
						matches[i, j] = matches[i - 1, j - 1];
					}
					else if (regex[j - 1] == MultipleCharacters)
					{
						matches[i, j] = matches[i, j - 2];

						if (regex[j - 2] == SingleCharacter || regex[j - 2] == text[i - 2])
						{
							matches[i, j] = matches[i, j] || matches[i - 1, j];
						}
					}
					else
					{
						matches[i, j] = false;
					}
				}
			}

			return matches[text.Length, regex.Length];
		}
	}
}